util/src/main/java/io/kubernetes/client/informer/cache/Cache.java

```diff
@@ -0,0 +1,467 @@
+package io.kubernetes.client.informer.cache;
+
+import com.google.common.base.Strings;
+import io.kubernetes.client.informer.cache.function.IndexFunc;
+import io.kubernetes.client.informer.cache.function.KeyFunc;
+import io.kubernetes.client.informer.exception.BadObjectException;
+import io.kubernetes.client.models.V1ObjectMeta;
+import io.kubernetes.client.util.Reflect;
+import io.kubernetes.client.util.exception.ObjectMetaReflectException;
+import java.util.*;
+import java.util.concurrent.locks.ReentrantLock;
+
+/** Cache is java a portable for k/client-go's ThreadSafeStore */
+public class Cache<ApiType> implements Indexer<ApiType> {
+
+  /** keyFunc defines how to map objects into indices */
+  private KeyFunc<ApiType> keyFunc;
+
+  /** NAMESPACE_INDEX is the default index function for caching objects */
+  public static final String NAMESPACE_INDEX = "namespace";
+
+  /** indexers stores index functions by their names */
+  private Map<String, IndexFunc> indexers;
+
+  /** items stores object instances */
+  private Map<String, ApiType> items;
+
+  /** indices stores objects' keys by their indices */
+  private Map<String, Map<String, Set<String>>> indices = new HashMap<>();
+
+  /** lock provides thread-safety */
+  private ReentrantLock lock = new ReentrantLock();
```

Comment:
Do you really need this? I would just use `synchronized`

Reply:
mmm good question. technically, it would be fine to replace the lock instance by `synchronized`. wdyt if we do it in a follow-up? it's safer to keep the code on a same basis w/ its go code implementation (additionally it's proved to be working in our environment) as its first landing to this repo. i added a TODO note here.